bin/light-base/src/runtime_service.rs

```diff
@@ -593,98 +208,102 @@ impl<TPlat: Platform> RuntimeService<TPlat> {
         // This can take a long time.
         let mut guarded_lock = loop {
-            let mut guarded_lock = guarded.lock().await;
+            let guarded_lock = guarded.lock().await;
 
             match &guarded_lock.tree {
                 GuardedInner::FinalizedBlockRuntimeKnown { .. } => break guarded_lock,
-                GuardedInner::FinalizedBlockRuntimeUnknown { .. } => {
-                    let (tx, mut rx) = mpsc::channel(0);
-                    let subscription_id = guarded_lock.next_subscription_id;
-                    guarded_lock.next_subscription_id += 1;
-                    guarded_lock
-                        .all_blocks_subscriptions
-                        .insert(subscription_id, (tx, 8));
+                GuardedInner::FinalizedBlockRuntimeUnknown { when_known, .. } => {
+                    let wait_fut = when_known.listen();
                     drop(guarded_lock);
-                    let _ = rx.next().await;
+                    wait_fut.await;
                 }
             }
         };
         let mut guarded_lock = &mut *guarded_lock;
 
+        // Extract the components of the `FinalizedBlockRuntimeKnown`. We are guaranteed by the
+        // block above to be in this state.
+        let (tree, finalized_block, pinned_blocks, all_blocks_subscriptions) =
+            match &mut guarded_lock.tree {
+                GuardedInner::FinalizedBlockRuntimeKnown {
+                    tree,
+                    finalized_block,
+                    pinned_blocks,
+                    all_blocks_subscriptions,
+                } => (
+                    tree,
+                    finalized_block,
+                    pinned_blocks,
+                    all_blocks_subscriptions,
+                ),
+                _ => unreachable!(),
+            };
+
         let (mut tx, new_blocks_channel) = mpsc::channel(buffer_size);
         let subscription_id = guarded_lock.next_subscription_id;
         guarded_lock.next_subscription_id += 1;
 
-        let non_finalized_blocks_ancestry_order: Vec<_> = match &guarded_lock.tree {
-            GuardedInner::FinalizedBlockRuntimeKnown {
-                tree,
-                finalized_block,
-            } => {
-                let decoded_finalized_block =
-                    header::decode(&finalized_block.scale_encoded_header).unwrap();
+        let decoded_finalized_block =
+            header::decode(&finalized_block.scale_encoded_header).unwrap();
+        pinned_blocks.insert(
+            (subscription_id, finalized_block.hash),
+            (
+                tree.finalized_async_user_data().clone(),
+                *decoded_finalized_block.state_root,
+                decoded_finalized_block.number,
+            ),
+        );
+
+        let non_finalized_blocks_ancestry_order: Vec<_> = tree
+            .input_iter_ancestry_order()
+            .filter_map(|block| {
```

Comment:
why do we need `filter` if we always return `Some`?

Reply:
It's not always `Some`, the line just below the `filter_map` has a `?`